epochStart/bootstrap/metaStorageHandler.go

```diff
@@ -163,6 +177,139 @@ func (msh *metaStorageHandler) SaveDataToStorage(components *ComponentsNeededFor
+	epochStartMetaBootstrapInfo bootstrapStorage.BootstrapHeaderInfo,
+	syncedHeaders map[string]data.HeaderHandler,
+) ([]bootstrapStorage.BootstrapHeaderInfo, error) {
+	var lastSelfNotarizedHeaders []bootstrapStorage.BootstrapHeaderInfo
+	if !epochStartMeta.IsHeaderV3() {
+		return []bootstrapStorage.BootstrapHeaderInfo{
+			epochStartMetaBootstrapInfo,
+		}, nil
+	}
+
+	for _, epochStartData := range epochStartMeta.GetEpochStartHandler().GetLastFinalizedHeaderHandlers() {
+
+		bootstrapHdrInfo, err := msh.getLastNotarizedBootstrapInfoForEpochStartData(epochStartData, syncedHeaders)
+		if err != nil {
+			return nil, err
+		}
+
+		lastSelfNotarizedHeaders = append(lastSelfNotarizedHeaders, bootstrapHdrInfo)
+	}
+
+	bootstrapHdrInfoMeta, err := msh.getLastMetaBootstrapInfo(epochStartMeta, syncedHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	lastSelfNotarizedHeaders = append(lastSelfNotarizedHeaders, bootstrapHdrInfoMeta)
+
+	return lastSelfNotarizedHeaders, nil
+}
+
+func (msh *metaStorageHandler) getLastMetaBootstrapInfo(
+	epochStartMeta data.MetaHeaderHandler,
+	syncedHeaders map[string]data.HeaderHandler,
+) (bootstrapStorage.BootstrapHeaderInfo, error) {
+	lastExecRes, err := common.GetLastBaseExecutionResultHandler(epochStartMeta)
+	if err != nil {
+		return bootstrapStorage.BootstrapHeaderInfo{}, err
+	}
+
+	lastExecMetaHeader, ok := syncedHeaders[string(lastExecRes.GetHeaderHash())]
+	if !ok {
+		return bootstrapStorage.BootstrapHeaderInfo{}, epochStart.ErrMissingHeader
+	}
+
+	bootstrapHdrInfoMeta := bootstrapStorage.BootstrapHeaderInfo{
+		ShardId: core.MetachainShardId,
+		Epoch:   lastExecMetaHeader.GetEpoch(),
+		Nonce:   lastExecMetaHeader.GetNonce(),
+		Hash:    lastExecRes.GetHeaderHash(),
+	}
+
+	return bootstrapHdrInfoMeta, nil
+}
+
+func (msh *metaStorageHandler) getLastNotarizedBootstrapInfoForEpochStartData(
+	epochStartData data.EpochStartShardDataHandler,
+	syncedHeaders map[string]data.HeaderHandler,
+) (bootstrapStorage.BootstrapHeaderInfo, error) {
+	lastFinishedMetaBlockHash := epochStartData.GetLastFinishedMetaBlock()
+	lastFinishedMetaBlock, ok := syncedHeaders[string(lastFinishedMetaBlockHash)]
+	if !ok {
+		return bootstrapStorage.BootstrapHeaderInfo{}, epochStart.ErrMissingHeader
+	}
+
+	shardHeaderHash := epochStartData.GetHeaderHash()
+	shardHeader, ok := syncedHeaders[string(shardHeaderHash)]
+	if !ok {
+		return bootstrapStorage.BootstrapHeaderInfo{}, epochStart.ErrMissingHeader
+	}
+
+	shardHeaderHandler, ok := shardHeader.(data.ShardHeaderHandler)
+	if !ok {
+		return bootstrapStorage.BootstrapHeaderInfo{}, epochStart.ErrWrongTypeAssertion
+	}
+
+	if len(shardHeaderHandler.GetMetaBlockHashes()) <= 0 {
+		return bootstrapStorage.BootstrapHeaderInfo{
+			ShardId: epochStartData.GetShardID(),
+			Epoch:   lastFinishedMetaBlock.GetEpoch(),
+			Nonce:   lastFinishedMetaBlock.GetNonce(),
+			Hash:    lastFinishedMetaBlockHash,
+		}, nil
+	}
+
+	metaHash := shardHeaderHandler.GetMetaBlockHashes()[0]
```

Comment:
as discussed, the metaHash should be the last referenced metablock on the shardHeader.LastExecutinResult 
if no Meta referenced there, then search on the previous and so forth until one is found.

Reply:
updated